### PR TITLE
[eclipse/xtext#1679] Refactor more Xtend to Java

### PR DESCRIPTION
**Pull request by szarnekow**
[eclipse/xtext#1679] Refactor more Xtend to Java

### CODE REVIEW

File: org.eclipse.xtext/src/org/eclipse/xtext/naming/DefaultCopyQualifiedNameService.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2018 itemis AG (http://www.itemis.eu) and others.
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/naming/DefaultCopyQualifiedNameService.java]: Year 2018, 2020

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+		public PortableFragmentDescription(EClass descriptionEClass, QualifiedName descriptionQualifiedName,
+				String descriptionRelativeFragment) {
+			this.descriptionEClass = descriptionEClass;
+			this.descriptionQualifiedName = descriptionQualifiedName;
+			this.descriptionRelativeFragment = descriptionRelativeFragment;
+		}
+
+		public EClass getDescriptionEClass() {
+			return this.descriptionEClass;
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: `this.` not needed here, and on lines 77, 81.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+	 * 
+	 * @return the EObject for the given portableURIFragment
+	 */
+	public EObject resolve(StorageAwareResource resource, String portableFragment) {
+		PortableURIs.PortableFragmentDescription desc = this.fromFragmentString(portableFragment);
+		EReference mock = EcoreFactory.eINSTANCE.createEReference();
+		mock.setEType(desc.descriptionEClass);
+		IScope scope = this.globalScopeProvider.getScope(resource, mock, Predicates.<IEObjectDescription>alwaysTrue());
+		Optional<IEObjectDescription> description = Streams.stream(scope.getElements(desc.descriptionQualifiedName)).findFirst();
+		return description.map(d->{
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Please format code.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+			EObject container = EcoreUtil.resolve(d.getEObjectOrProxy(), resource);
+			return getEObject(container, desc.descriptionRelativeFragment);
+		}).orElse(null);
+	}
+
+	/**
+	 * Creates and returns a portable URI from the given resource to the targetURI. Returns <code>null</code> is no
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Returns <code>null</code> *if* ...

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+		EObject object = sourceResource.getResourceSet().getEObject(targetURI, false);
+		// if it points to some registered ecore, there's no resourceSet and the result is not portable
+		if (object == null || object.eResource().getResourceSet() != null) {
+			return toPortableURI(sourceResource, object);
+		}
+		return null;
+	}
+
+	/**
+	 * Creates and returns a portable URI from the global scope. Returns <code>null</code> is no portable URI can be
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: PS comments above.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	 *            the target object that should be resolvable by the created portable URI
+	 * 
+	 * @return a portable URI or <code>null</code>
+	 */
+	public URI toPortableURI(StorageAwareResource sourceResource, EObject targetObject) {
+		if (targetObject == null || targetObject.eIsProxy()) {
+			return sourceResource.getURI().appendFragment(StorageAwareResource.UNRESOLVABLE_FRAGMENT);
+		}
+		String portableFragment = getPortableURIFragment(targetObject);
+		if ((portableFragment != null)) {
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Excessive ((

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+	}
+
+	/**
+	 * @return a portable URI fragment, or <code>null</code> if the give EObject isn't itself or is not contained in an
+	 *         exported EObjectDescription
+	 */
+	protected String getPortableURIFragment(EObject obj) {
+		IResourceDescriptions descriptions = resourceDescriptionsProvider.getResourceDescriptions(obj.eResource());
+		IResourceDescription desc = descriptions.getResourceDescription(obj.eResource().getURI());
+		if ((desc == null)) {
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Excessive ((

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+			return null;
+		}
+		return FluentIterable.from(desc.getExportedObjects()).filter(description->{
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Did you purposefully use `FluentIterable` here?

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageProviderAdapter.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+public class ResourceStorageProviderAdapter extends AdapterImpl {
+	@Override
+	public boolean isAdapterForType(Object type) {
+		return ResourceStorageProviderAdapter.class.equals(type);
+	}
+
+	public ResourceStorageLoadable getResourceStorageLoadable(StorageAwareResource resource) {
+		return null;
+	}
+}
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageProviderAdapter.java]: NL

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2014 itemis AG (http://www.itemis.eu) and others.
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java]: 2014, 2020

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+				return (SourceLevelURIsAdapter) adapter;
+			}
+		}
+		return null;
+	}
+	
+	private Set<URI> sourceLevelURIs;
+
+	public Set<URI> getSourceLevelURIs() {
+		return Collections.unmodifiableSet(this.sourceLevelURIs);
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java]: `this.` not needed.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2014, 2017 itemis AG (http://www.itemis.eu) and others.
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: 2014, 2020

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+import org.eclipse.xtext.util.internal.Stopwatches;
+
+import com.google.inject.Inject;
+
+/**
+ * A resource implementation that can load itself from ResourceStorage.
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ */
+@SuppressWarnings("all")
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: Please remove if there are no warnings in the code.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+		if (!isLoaded && !isLoading && resourceStorageFacade != null
+				&& resourceStorageFacade.shouldLoadFromStorage(this)) {
+			if (StorageAwareResource.LOG.isDebugEnabled()) {
+				LOG.debug("Loading " + getURI() + " from storage.");
+			}
+			try {
+				ResourceStorageLoadable in = resourceStorageFacade.getOrCreateResourceStorageLoadable(this);
+				loadFromStorage(in);
+				return;
+			} catch (IOException e) {
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: Code formatting inside the `catch` block seems to be broke.

**Comment by nbhusare** [comment]
Minot nits and LGTM

**Comment by cdietrich** [comment]
@szarnekow please also check for the test failures

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java (hunk 8 above)
Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: In fact I did, but it was trivial to use Streams.stream instead.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java (hunk 11 above)
Comment (szarnekow) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SourceLevelURIsAdapter.java]: I'll still return the unmodifiableSet though.

**Comment by szarnekow** [comment]
Thank you for the review, @nbhusare - I updated the commit.

**Comment by szarnekow** [comment]
> @szarnekow please also check for the test failures

Done. We used to throw IllegalStateException (from a stateless service 🤔 ) - added JavaDoc and fixed test to expect IllegalArgumentException

**Comment by cdietrich** [comment]
@tivervac @nbhusare please reexamine

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+				}
+				unload();
+			}
+		}
+		super.load(options);
+	}
+
+	public void loadFromStorage(ResourceStorageLoadable storageInputStream) throws IOException {
+		if (storageInputStream == null) {
+			throw new NullPointerException("storageInputStream");
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: You might like to change the detailed message to `the parameter storageInputStream must not be null` or something better.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+	protected String getPortableURIFragment(EObject obj) {
+		IResourceDescriptions descriptions = resourceDescriptionsProvider.getResourceDescriptions(obj.eResource());
+		IResourceDescription desc = descriptions.getResourceDescription(obj.eResource().getURI());
+		if (desc == null) {
+			return null;
+		}
+		return Streams.stream(desc.getExportedObjects()).filter(description -> {
+			EObject possibleContainer = EcoreUtil.resolve(description.getEObjectOrProxy(), obj.eResource());
+			return EcoreUtil.isAncestor(obj, possibleContainer);
+
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Please remove line.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+	 * 
+	 * @param resource
+	 *            the resource from which global scope to look up the EObject
+	 * @param portableFragment
+	 *            the portable fragment pointing to the to be resolved EObject
+	 * 
+	 * @return the EObject for the given portableURIFragment
+	 */
+	public EObject resolve(StorageAwareResource resource, String portableFragment) {
+		PortableURIs.PortableFragmentDescription desc = this.fromFragmentString(portableFragment);
```
Comment (nbhusare) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/PortableURIs.java]: Please remove `this` on lines 158, and 161.

File: org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/resource/persistence/PortableURIsTest.java [hunk 18]
```diff
@@ -22,6 +22,7 @@
 import org.eclipse.xtext.linking.langATestLanguage.Type;
```
Comment (nbhusare) [inline org.eclipse.xtext.tests/xtend-gen/org/eclipse/xtext/resource/persistence/PortableURIsTest.java]: Please add 2020 to the Copyright if you think its required.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/persistence/PortableURIsTest.xtend [hunk 19]
```diff
@@ -84,7 +84,7 @@ class PortableURIsTest extends AbstractXtextTests {
 		try {
```
Comment (nbhusare) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/resource/persistence/PortableURIsTest.xtend]: Please add 2020 to the Copyright if you think its required.

**Comment by nbhusare** [comment]
Minor nits and LGTM

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+	private boolean isLoadedFromStorage = false;
+
+	private IResourceDescription resourceDescription = null;
+
+	@Override
+	public void load(Map<?, ?> options) throws IOException {
+		if (!isLoaded && !isLoading && resourceStorageFacade != null
+				&& resourceStorageFacade.shouldLoadFromStorage(this)) {
+			if (StorageAwareResource.LOG.isDebugEnabled()) {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/StorageAwareResource.java]: Isn't this just done in LOG.debug?

**Pull request merged**